src/core/linalg/src/dense/4C_linalg_tensor_operations.hpp

```diff
@@ -0,0 +1,416 @@
+    using value_type = Tensor::value_type;
+    constexpr std::size_t m = Tensor::template extent<0>();
+    constexpr std::size_t n = Tensor::template extent<1>();
+
+    Core::LinAlg::Tensor<value_type, n, m> dest;
+    DenseFunctions::update_t<value_type, n, m>(dest.data(), A.data());
+    return dest;
+  }
+
+  template <typename TensorLeft, typename TensorRight>
+    requires(Rank2TensorConcept<TensorLeft> && Rank1TensorConcept<TensorRight> &&
+             TensorLeft::template extent<1>() == TensorRight::template extent<0>())
+  constexpr auto dot(const TensorLeft& A, const TensorRight& b)
+  {
+    using value_type = decltype(std::declval<typename TensorLeft::value_type>() *
+                                std::declval<typename TensorRight::value_type>());
+    constexpr std::size_t m = TensorLeft::template extent<0>();
+    constexpr std::size_t n = TensorLeft::template extent<1>();
+
+    Tensor<value_type, m> dest;
+    DenseFunctions::multiply<value_type, m, n, 1>(dest.data(), A.data(), b.data());
+    return dest;
+  }
+
+  template <typename TensorLeft, typename TensorRight>
+    requires(Rank1TensorConcept<TensorLeft> && Rank2TensorConcept<TensorRight> &&
+             TensorLeft::template extent<0>() == TensorRight::template extent<0>())
+  constexpr auto dot(const TensorLeft& a, const TensorRight& B)
+  {
+    using value_type = decltype(std::declval<typename TensorLeft::value_type>() *
+                                std::declval<typename TensorRight::value_type>());
+    constexpr std::size_t m = TensorRight::template extent<0>();
+    constexpr std::size_t n = TensorRight::template extent<1>();
+
+    Tensor<value_type, n> dest;
+    DenseFunctions::multiply<value_type, 1, m, n>(dest.data(), a.data(), B.data());
+    return dest;
+  }
+
+  template <typename TensorLeft, typename TensorRight>
+    requires(Rank2TensorConcept<TensorLeft> && Rank2TensorConcept<TensorRight> &&
+             TensorLeft::template extent<1>() == TensorRight::template extent<0>())
+  constexpr auto dot(const TensorLeft& A, const TensorRight& B)
+  {
+    using value_type = decltype(std::declval<typename TensorLeft::value_type>() *
+                                std::declval<typename TensorRight::value_type>());
+    constexpr std::size_t m = TensorLeft::template extent<0>();
+    constexpr std::size_t k = TensorLeft::template extent<1>();
+    constexpr std::size_t n = TensorRight::template extent<1>();
+
+    Core::LinAlg::Tensor<value_type, n, m> dest;
+    DenseFunctions::multiply<value_type, m, k, n>(dest.data(), A.data(), B.data());
+    return dest;
+  }
+
+  template <typename TensorLeft, typename TensorRight>
+    requires(Rank2TensorConcept<TensorLeft> && Rank2TensorConcept<TensorRight> &&
+             TensorLeft::template extent<0>() == TensorRight::template extent<0>() &&
+             TensorLeft::template extent<1>() == TensorRight::template extent<1>())
+  constexpr auto ddot(const TensorLeft& A, const TensorRight& B)
```

Comment:
These operations are all very nice! Could you add docs? And, again, I suggest to move the definition at the end of the header and just have the declaratiion with docs here.